src/Experiment/src/AssemblyBuilder.cs

```diff
@@ -1,27 +1,68 @@
-using System.Reflection;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Reflection.PortableExecutable;
 
 namespace System.Reflection.Emit.Experimental
 {
+
     public class AssemblyBuilder: System.Reflection.Assembly
     {
-        internal AssemblyBuilder() { }
+        private static readonly Guid s_guid = Guid.NewGuid();
+        private BlobBuilder _emptyBlob = new BlobBuilder();//For later
+        private AssemblyName? _assemblyName;
+        private MetadataBuilder _metadata = new MetadataBuilder();
+        private IDictionary<string,ModuleBuilder> _moduleStorage = new Dictionary<string, ModuleBuilder>();
+
+        public AssemblyBuilder() { }
+        private AssemblyBuilder(AssemblyName name) 
+        {
+            _assemblyName = name;
+        }
 
         public void Save(string assemblyFileName)
         {
-            //New method
+            if (assemblyFileName == null)
+            {
+                throw new ArgumentNullException();
+            }
+            if (_assemblyName==null||_assemblyName.Name==null)
+            {
+                throw new ArgumentNullException();
+            }
+            //Add assembly metadata
+            _metadata.AddAssembly(//Metadata is added for the new assembly - Current design - metdata generated only when Save method is called.
+               _metadata.GetOrAddString(value: _assemblyName.Name),
+               version: _assemblyName.Version ?? new Version(1, 0, 0, 0),
+               culture: (_assemblyName.CultureName==null) ?  default : _metadata.GetOrAddString(value: _assemblyName.CultureName),
+               publicKey: (_assemblyName.GetPublicKey() is byte[] publicKey) ? _metadata.GetOrAddBlob(value: publicKey) : default,
+               flags: (AssemblyFlags) _assemblyName.Flags,
+               hashAlgorithm: AssemblyHashAlgorithm.None);//It seems AssemblyName.HashAlgorithm is obslete so default value used.
+            //Add each module's medata
+            foreach (KeyValuePair<string, ModuleBuilder> entry in _moduleStorage)
+            {
+                entry.Value.AppendMetadata(_metadata);
+            }
+            using var peStream = new FileStream(assemblyFileName, FileMode.OpenOrCreate, FileAccess.ReadWrite);
+            var ilBuilder = new BlobBuilder();
+            WritePEImage(peStream, _metadata, ilBuilder);
         }
 
-        public static System.Reflection.Emit.AssemblyBuilder DefineDynamicAssembly(System.Reflection.AssemblyName name, System.Reflection.Emit.AssemblyBuilderAccess access) 
-        { 
-            throw new NotImplementedException(); 
+        public static System.Reflection.Emit.Experimental.AssemblyBuilder DefineDynamicAssembly(System.Reflection.AssemblyName name, System.Reflection.Emit.AssemblyBuilderAccess access)
+        {
+            if (name == null || name.Name == null)
+            {
+                throw new ArgumentNullException();
```

Comment:
```suggestion
                throw new ArgumentNullException(nameof(name));
```
So the message includes the parameter's name